frontend/public/declarations.d.ts

```diff
@@ -41,7 +41,7 @@ declare interface Window {
     GOOS: string;
     graphqlBaseURL: string;
   };
-  windowError?: boolean | string;
+  windowError?: PromiseRejectionEvent;
```

Comment:
This isn't always a promise rejection. It can be any error.

Reply:
Isn't [this](https://github.com/openshift/console/blob/b8325c2f9032b1e822e1f5f63fb5506394b4e515/frontend/public/components/app.jsx#L217) just trapping `onunhandledrejection`?

Reply:
No, it's also `window.onerror`

Reply:
Ok, so `window.onerror = (e Event | string) => ...`, either way we're trapping an Event | string and setting `windowError` to it.  It's not an `Error` with a stack trace.

Reply:
Ah, we should really be handling the `error`, which is passed, just not as the first arg.

https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror

We probably want to switch to `window.addEventListener('error')`

Reply:
k, I was considering that as well, thanks

Reply:
It would have to be something like:
```
window.addEventListener('error', e:ErrorEvent => {
    // Get the error properties from the error event object
    const { message, source, lineno, colno, error } = e; 
   // construct new Error from ^^ and assign to win.windowError ?
   //  or set win.windowError to e (ErrorEvent)?
});
```

Reply:
How about?

```ts
window.addEventListener('error', ({error}) => window.windowError = error);
```

The key thing is that we retain the original stack trace and any error details for debugging this failure in CI.

Reply:
When I use `window.onerror = (e : Event | string) => {` or `window.addEventListener('error', (e: ErrorEvent) => {` in app.js it doesn't trap the `PromiseRejectionEvent` thrown -that is only caught if I use `window.onerror = window.onunhandledrejection = (e) => {`
If I use `window.onerror = window.onunhandledrejection = (e) => {` then the `PromiseRejectionEvent` is caught in app.js; but it is an Event not an Error.

I'm surprised the `window.onerror` or ` `window.addEventListener('error',` don't catch the Error thrown by co-fetch.js's `validateStatus` which processes the fetch.response and generates any appropriate [Error](https://github.com/openshift/console/blob/cf7744e55f10bd7f05eafd57d8485254862faba8/frontend/public/co-fetch.js#L73).  If I could catch the Error thrown, then I could use it's `error.response`.

Reply:
Seems like something is dispatching the `PromiseRejectionEvent` when url is not found (my way of triggering it).  I thought throwing an Error would be the same as dispatching an ErrorEvent, but it doesn't seem so.

Reply:
You still need a `window.onunhandledrejection` handler.

Reply:
resolved, pls see latest code